src/api_client.js

```diff
@@ -140,24 +141,48 @@ export class ApiClient {
   }
 
   // Execute an API request against the Submarine API.
-  execute(method, data, context, callback) {
+  async execute(method, data, context, callback) {
     const url = getMethodUrl(this.environment, method, context);
     const http_method = getMethodHttpMethod(method);
     const queryParams = this.buildQueryParams(http_method, data);
     const payload = getMethodPayload(http_method, data);
+    let jwtToken = null;
+
+    let result = await this.fetchJWTToken();
+      
+    if (result.errors) {
+      callback && callback(null, result.errors);
+      return;
+    } else {
+      window.localStorage.setItem(context.customer_id, result.token);
+      jwtToken = result.token;
+    }
 
     return fetch(url + buildQueryString(queryParams), {
       method: http_method,
       mode: 'cors',
       headers: {
-        'Content-Type': 'application/json; charset=utf-8'
+        'Content-Type': 'application/json; charset=utf-8',
+        'Authorization': jwtToken
       },
       body: payload
     })
       .then(response => {
         response.json()
-          .then(json => {
-            if(json && json.errors) {
+          .then(async json => {
+
+            if (response.status === 401) {
+              let result = await this.fetchJWTToken();
```

Comment:
@appu4ever i'm not sure if this is correct.

I know gavin suggested a different error handling strategy than the one you and i spoke about, but this just got into an infinite loop if i set submarine to always return a 401.

If the new strategy is to get a fresh token every time, then if the call to submarine returns a 401, then it shouldn't retry (i.e. it shouldn't attempt to get another token)